jobserver/views/users.py

```diff
@@ -1,82 +1,101 @@
 from django.contrib import messages
+from django.contrib.auth import login
 from django.contrib.auth.decorators import login_required
-from django.contrib.auth.views import PasswordResetConfirmView
+from django.core.signing import BadSignature, SignatureExpired
 from django.shortcuts import redirect
-from django.template.response import TemplateResponse
 from django.utils.decorators import method_decorator
-from django.views.generic import FormView, UpdateView
+from django.views.generic import FormView, UpdateView, View
+from opentelemetry import trace
 
-from ..authorization import InteractiveReporter
-from ..emails import send_github_login_email, send_reset_password_email
-from ..forms import ResetPasswordForm
+from jobserver.authorization import InteractiveReporter, has_role
+
+from ..emails import send_github_login_email, send_login_email
+from ..forms import EmailLoginForm
 from ..models import User
 from ..utils import is_safe_path
 
 
-def login_view(request):
-    next_url = request.GET.get("next") or "/"
-    if not is_safe_path(next_url):
-        next_url = ""
+INTERNAL_LOGIN_SESSION_TOKEN = "_login_token"
 
-    if request.user.is_authenticated:
-        return redirect(next_url)
+logger = structlog.get_logger(__name__)
+tracer = trace.get_tracer(__name__)
+
+
+class Login(FormView):
+    form_class = EmailLoginForm
+    template_name = "login.html"
 
-    return TemplateResponse(request, "login.html", {"next_url": next_url})
+    def dispatch(self, request, *args, **kwargs):
+        self.next_url = self.get_next_url(request)
 
+        if request.user.is_authenticated:
+            return redirect(self.next_url)
 
-class ResetPassword(FormView):
-    form_class = ResetPasswordForm
-    template_name = "reset_password.html"
+        return super().dispatch(request, *args, **kwargs)
 
     def form_valid(self, form):
         user = User.objects.filter(email=form.cleaned_data["email"]).first()
-
-        # we have three possible states for the given email:
-        # * unknown user
-        # * GitHub user
-        # * email-only user
-        # we want to show the same outcome regardless of state so a bad actor
-        # can't work out work out who uses the service by email
         if user:
             if user.social_auth.exists():
-                # send GitHub users a link to the PSA entrypoint and explain
-                # that they don't have a password to reset
+                # we don't want to expose users email address to a bad actor
+                # via the login page form so we're emailing them with a link
+                # to the login page to click the right button.
+                # We can't email them with a link to the social auth entrypoint
+                # because it only accepts POSTs.
                 send_github_login_email(user)
             else:
-                send_reset_password_email(user)
+                self.request.session[INTERNAL_LOGIN_SESSION_TOKEN] = user.email
```

Comment:
I think we should refuse to send an email if they're not interactive users? Same validation as like the POST handler, basically